File: naive-bayes/naive-bayes-movie.py
# -*- coding: utf-8 -*-
"""
Created on Mon Dec 13 21:07:28 2021

@author: parkk
"""



# 나이가 20대이고 성별이 여자이며 직업이 IT 이고 결혼을 안했으며
# 이성친구가 없는 사람이 선택할 가능성이 높은 영화 장르는 ?  ( 파이썬으로, moive.csv )


import pandas as pd
movie = pd.read_csv("c:\\data\\movie.csv", encoding = 'euckr')
# print ( movie.isnull().sum() )     # 결측치확인

x = movie.iloc[ : , :5 ]            # 정답을 뺀 데이터 생성
y = movie.iloc [  : , 5]            # 정답 데이터 생성

movie2 = pd.get_dummies(x)  #  5개 컬럼에서 18개로 늘어남 , 숫자 데이터로 변환

from sklearn.preprocessing import MinMaxScaler         # 데이터를 정규화 합니다.

scaler = MinMaxScaler()
scaler.fit(movie2)
movie2_scaled = scaler.transform(movie2)

y = y.to_numpy()           # y를 numpy로 만들어줌

from sklearn.model_selection import train_test_split       # 훈련데이터와 테스트 데이터 분리

x_train, x_test, y_train, y_test = train_test_split ( movie2_scaled, y , test_size = 0.2, random_state = 1 )

from sklearn.naive_bayes import BernoulliNB      # 나이브 베이즈 모델 생성
 
model = BernoulliNB()

model.fit(x_train, y_train)          # 모델 훈련

result = model.predict(x_test)          # 예측

print ( (sum( result == y_test )) / (len(y_test)) * 100)          # 87.5

# 성능 향상 & var_smoothing 값 찾기

x_val = list(range(1,1000))
y_val = []

for i in range(1, 1000):
    # from sklearn.naive_bayes import BernoulliNB
    from sklearn.naive_bayes import GaussianNB
    # from sklearn.naive_bayes import MultinomialNB

    model2 = GaussianNB( var_smoothing = (i /1000 ) ) 
    model2.fit(x_train, y_train)
    result2 = model2.predict(x_test)

    from sklearn.metrics import confusion_matrix
    a = sum( result2 == y_test ) / (len(y_test)) * 100
    y_val.append(a)    
                              
plot_dict = { 'i' : x_val, '확률' : y_val }
plot_dict2 = pd.DataFrame(plot_dict)
print(plot_dict2.plot(kind = 'line', x = 'i', y = '확률'))


# i = 599 이상일때 ( var_smoothing가 0.599 이상 ) 확률이 100이다 (a == 100) 

######  예측할 데이터 생성

print(movie2.columns)       # 숫자로 변형한 데이터의 컬럼 확인
						
# ['나이_10대', '나이_20대', '나이_30대', '나이_40대', '성별_남', '성별_여', '직업_IT',
# '직업_디자이너', '직업_무직', '직업_언론', '직업_영업', '직업_자영업', '직업_학생', '직업_홍보/마케팅',
 # '결혼여부_NO', '결혼여부_YES', '이성친구_NO', '이성친구_YES']

exp_data = {}
temp_list = [0]*len(movie2.columns)

for i,k in zip(movie2.columns, temp_list ):
    exp_data[i] = k 

exp_data['나이_20대'] = 1
exp_data['직업_IT'] = 1
exp_data['성별_여'] = 1
exp_data['결혼여부_NO'] = 1
exp_data['이성친구_NO'] = 1

exp_data2 = pd.DataFrame ( exp_data, index = [0] )   

# 결과 예측
result2 = model2.predict(exp_data2.to_numpy())          # 예측
print(result2)                   # 정답 : 로맨틱

File: naive-bayes/naive-bayes-movie.R
# 직업과 성별과 결혼 유무등의 데이터를 가지고 영화장르를 예측하는데
# 나이브 베이즈 모델을 생성하시오 ! ( 데이터 셋 : movie.csv )

# 1. 데이터를 로드합니다.

movie <- read.csv("movie.csv", stringsAsFactor = TRUE)
str(movie)         # all factor

# 맨앞에 있는 type 이 라벨(정답) 입니다.

table(movie$장르)

#  SF   공포 로맨틱   무협 스릴러   액션 코미디 
#  5      5      9      5      5      5      5 


prop.table(table( movie$장르 ))                    #  비율을 알수있음.

#      SF      공포    로맨틱      무협    스릴러      액션    코미디 
# 0.1282051 0.1282051 0.2307692 0.1282051 0.1282051 0.1282051 0.1282051

dim(mush)                 # [1] [1] 39  6 , 전체 건수가 어떻게 되는지 확인

# 2. 결측치를 확인합니다.

colSums(is.na(movie))                    # 모두 0 이다.

# 3. 이상치를 확인합니다.

# 명목형 데이터여서 이상치를 확인할 필요가 없습니다.

# 4. 명목형 데이터가 있는지 확인합니다.

# 전부 명목형 데이터 입니다.

# 5. 데이터를 정규화 합니다.

# 전부 명목형 데이터 이므로 정규화 작업도 필요하지 않습니다.

# 6. 훈련 데이터와 테스트 데이터를 분리합니다.

# 데이터 shuffle 과 데이터 분리를 효율적이면서도 편하게 수행할 수 있는 패키지를 이용해서 분리해보겠습니다.

install.packages("caret")
library(caret)

set.seed(1)
k <- createDataPartition( movie$장르, p = 0.8 , list = F )   # 훈련데이터 80 %, 테스트 20 %  / sample 사용하지 않고 쉽게 만듦
# 또한, 리스트 형태로 만들지 말아라
k           # 훈련데이터의 index가 나옴

train_data <- movie[ k, ]
test_data <- movie[ -k,  ]

dim(train_data)           # [1] 32  6
dim(test_data)           # [1] 7 6

prop.table( table(train_data$장르) ) 

#    SF   공포 로맨틱   무협 스릴러   액션 코미디 
#  0.125  0.125  0.250  0.125  0.125  0.125  0.125 

prop.table( table(test_data$장르) )

#        SF      공포    로맨틱      무협    스릴러      액션    코미디 
# 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571 0.1428571

# ※ 훈련데이터와 테스트 데이터의 독버섯과 정상버섯이 거의 50:50으로 균등하게 분포되어있습니다.
# sample에 비해서 데이터가 더 훈련하고 테스트하기 쉽게 잘 나눠져서 좋다.


# 7. 나이브 베이즈 모델을 생성합니다.

install.packages("e1071")
library(e1071)

model <- naiveBayes( type~ . , data= train_data )        # type 는 라벨 ,  '.'은 라벨외의 모든컬럼을 뜻함
model


# cap_shape = 버섯 cap의 모양
# cap_surface = 버섯 cap의 표면,, 등등

# 버섯 데이터로 빈도표를 만들고서 우도표를 생성했다.

# 8. 훈련 데이터와 라벨( 정답 )으로 모델을 훈련시킵니다.

# 7번에서 다 수행했습니다.


# 9. 훈련된 모델로 테스트 데이터를 예측합니다.

result <- predict( model, test_data[  , -1] )              # 정답 컬럼 빼고

# 테스트 데이터의 정답을 제외하고 예측합니다.
result

# 10. 모델의 성능을 평가합니다.

sum ( (result == test_data[, 1]) / length(test_data[ , 1]) )         # 

# 어제는 유방암 데이터가 전부 숫자여서 knn 알고리즘을 이용해서 기계학습 시켰고
# 오늘은 독버섯 데이터가 전부 명목형이어서 naivebayes 를 이용해서 기계학습 시켰습니다.


# 11. 모델의 성능을 높입니다.

model2 <- naiveBayes( type~ . , data = train_data, laplace = 0.0001 )

result2 <- predict( model2, test_data[  , -1] )

sum ( (result2 == test_data[, 1]) / length(test_data[ , 1]) )          # 

